### ISSUE TITLE
Latency steadily increases to 4 seconds with an InStream on a non-raw device

### ISSUE THREAD

**Issue by tzaeru** [Timmmm/soundio-rs#2]
Hi,

When using InStream, it seems that if I open the stream on a non-raw device, latency as given by the stream.get_latency() function in the stream callback steadily increases until it plateaus at 4 seconds. When I open a device as raw, this doesn't happen and latency stays at an acceptable level.

It would be very useful if I could open devices as non-raw, since it'd need to do less format management and wouldn't need to take exclusive control of a device.

Do you think this might be related to soundio-rs or might it be a bug in libsoundio?

**Comment by Timmmm** [comment]
Highly unlikely to be a soundio-rs issue because it mostly is just a wrapper around libsoundio. What value do you give as the requested latency when you open the stream?

Also - and I should really put this in the readme - I have kind of abandoned this project because I couldn't work out a way to separate lifetimes and borrowing, which makes things really awkward when you want to open multiple devices or streams.

In other words, if you open a device, it must live until at least as long as the library object. However the only way to make that happen in Rust (AFAIK, I am a noob) is to make it borrow the library object. But that means the library object is then borrowed so you can't do anything else with it, like opening another device.

There's probably a solution but it got complicated and I gave up.

**Comment by tzaeru** [comment]
I've tried with various latency values when opening the stream, from very low (0.05s) to pretty high (a few seconds). I'll try to see if google out something similar happening with libsoundio in C!

For the project being abandoned - well, it's less abandoned than the other Rust wrapper for libsoundio! I'm now using this with [rust-audio-analysis](https://github.com/tzaeru/rust-audio-analysis), which is used in production at where I work, so maybe I can see if I can do something to support it a little bit?

Without having actually looked into your code yet, I'd imagine that using Rc<Cell<T>> for the library object would be a solution. This would mean that the validity of references to the library object are checked at runtime.

**Comment by tzaeru** [comment]
I seem to be unable to replicate this on libsoundio's own examples.

It only happens on soundio-rs when using high latencies.

This is the kind of log I get for a software_latency of 0.1:

```
Latency: Ok(0.07)
Channel count: 2
Frame count: 441
Latency: Ok(0.11)
Channel count: 2
Frame count: 441
Latency: Ok(0.15)
Channel count: 2
Frame count: 441
Latency: Ok(0.19)
Channel count: 2
Frame count: 441
Latency: Ok(0.24)
Channel count: 2
Frame count: 441
Latency: Ok(0.28)
Channel count: 2
Frame count: 882
Latency: Ok(0.31)
Channel count: 2
Frame count: 882
Latency: Ok(0.34)
Channel count: 2
Frame count: 882
Latency: Ok(0.37)
Channel count: 2
Frame count: 441
```

When frame count is minimum on every frame (441), latency doesn't pile up.
